### ISSUE TITLE
[New Concept Docs]: `with` Statement

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3086]
This issue describes how to implement the `with` (keyword)  **concept docs**.
You can find the related **concept exercise** issue [here](https://github.com/exercism/python/issues/3087)

If you have not yet contributed to concept documents, this issue will require some upfront reading to give you the needed background knowledge.

## ✅   Getting started

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. 

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concepts:**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 

<br>

## 🎯   Goal

These concept docs are meant to teach an understanding/use of the `with` keyword and  general `context-handling`  in Python. 

<br>

## 💡   Learning objectives

- What the `with` keyword *is* and what it does in Python
  - context-handling & scope
  - setup and teardown processes
  - memory and resource allocation
- How to use the `with` keyword
- Most common use cases for managing resources with the `with` keyword
  - file opening/closing/processing
  - database connectivity
  - error handling
  - logging
  - testing
  - thread management (`threading.Lock()`)
- Nested `with` statements, and how to use them
- Gotchas and drawbacks of using `with`

<br>

## 🚫  Out of scope

<details>
<summary><b>Concepts and Subjects that are Out of Scope</b></summary>
<br>

- `classes`  & `class customization` 
- `class-inheritance` 
- `comprehensions` 
- `comprehensions` in `lambdas`
- `coroutiens` 
- `decorators` 
- `error-handling` beyond examples (_error handling will have its own detailed exercise_)
- `raising errors` beyond examples (_error raising will have its own detailed exercise_)
- `functions` and `higher-order functions`  beyond what might be needed to work with `with`
-  `functools` and related `map()`, `filter()` and `functools.reduce()`
- `generators` beyond what might be needed to work with `with` (_they have their own exercise_ )
- `lambdas` 
-  using an `assignment expression` or "walrus" operator (`:=`) 
- `enums`
- `testing` and `assert` (_these will have their own detailed exercise_)
-  The `csv` module (_this will get its own exercise_)

</details>

<br>

## 🤔   Concepts

<details>
<summary><b>Concepts and Related Concepts this Concept Covers</b></summary>
<br>

- `with` statement
- `context-management`
- `compound statements`

</details>

<br>

## ↩️   Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  

<details>
<summary>Prereqs</summary>
<br>

- `basics`
- `bools`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `functional-tools`
- `generators`
- `higher-order-functions`
- `iteration`
- `lists`
- `list-methods`
- `loops`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>  

## 📚   Resources to refer to

<details>
<summary><b>Resources</b></summary>
<br>

-[Sadique Ali Koothumadan: Understanding Python's `with` statement](https://sadique.io/blog/2012/07/09/understanding-pythons-with-statement/)
- [Real Python: Context Managers and Python's `with` statement](https://realpython.com/python-with-statement/)
-  [Dan Bader: Context Managers and the `with` statement in Python](https://dbader.org/blog/python-context-managers-and-with-statement)
- [Python Docs: Compound Statements](https://docs.python.org/3/reference/compound_stmts.html)
- [Python Docs: With Statement](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)
  - [PEP 0343: The `with` Statement](https://www.python.org/dev/peps/pep-0343/)
  - [PEP 0343:  Examples](https://www.python.org/dev/peps/pep-0343/#examples)
- [Python Docs:  Context Manager Types](https://docs.python.org/3/library/stdtypes.html#context-manager-types)

</details>
<details>
<summary><b>Additional Articles</b></summary>

- [Python Morsels: Context Handlers](https://www.pythonmorsels.com/context-managers/)
- [David Beazley at (Unenix.org) Python: With That - 5 Easy Context Managers ](https://www.usenix.org/system/files/login/articles/12_beazley-online.pdf)
- [Jeff Knupp: Python with Context Managers](https://jeffknupp.com/blog/2016/03/07/python-with-context-managers/)

</details>

<br>

## 📁   Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts)

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  -  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  -  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

* ### Concept `about.md`

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  * This can also be a summary/paraphrase of the  `about.md` document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the associated concept or practice exercises.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Concept `.meta/config.json` Entries
 
  For more information, see [Concept `.meta/config.json`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-metaconfigjson)

  * This file is likely already stubbed out.  Remember to add a concept **blurb** of less than 350 characters.  Please also add your GitHub username to the "authors" array, and any contributor GitHub usernames to the "contributors" array.

<br>
</td>
</table>
</details>

## 🎶  Implementation Notes

- Example code  should **only use syntax & concepts introduced within these docs or one of the prerequisite concept exercises or documents.** Where possible, please use REPL formatting, unless you are demonstrating pseudo code or a long code block.
  Please **do not use** syntax not previously covered in prerequisite topics or exercises.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

## 🆘   Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.
